File: sdk/python/lbro_sdk.py
"""LBRO Python SDK.

Lightweight SDK for sending security events to LBRO from Python applications.

Usage:
    from lbro_sdk import LBROClient, SecurityEvent

    client = LBROClient(
        api_key="proj_your_project_api_key",
        base_url="https://your-lbro-instance.example.com",
    )

    client.send_event(
        event_type="auth_failure",
        severity="high",
        source_ip="203.0.113.42",
        message="Failed SSH login attempt",
        payload={"username": "root", "attempts": 5},
    )

    # Batch send
    events = [
        SecurityEvent(event_type="sql_injection", severity="critical",
                      source_ip="192.0.2.1", message="SQL injection in /api/users"),
        SecurityEvent(event_type="xss", severity="medium",
                      source_ip="192.0.2.2", message="XSS in search field"),
    ]
    client.send_events(events)
"""
from __future__ import annotations

import json
import logging
import time
from dataclasses import dataclass, field, asdict
from datetime import datetime, timezone
from typing import Any
from urllib.error import HTTPError, URLError
from urllib.request import Request as URLRequest, urlopen

__version__ = "1.0.0"
__all__ = ["LBROClient", "SecurityEvent", "LBROError", "LBROAuthError", "LBROValidationError"]

log = logging.getLogger("lbro_sdk")

# Allowed event types
EVENT_TYPES = {
    "auth_failure", "sql_injection", "xss", "brute_force", "port_scan",
    "suspicious_request", "system_log", "application_log", "nginx_log",
    "apache_log", "firewall_event", "windows_event", "linux_audit", "custom",
}
SEVERITIES = {"critical", "high", "medium", "low", "info"}


class LBROError(Exception):
    """Base LBRO SDK error."""


class LBROAuthError(LBROError):
    """Invalid or expired project API key."""


class LBROValidationError(LBROError):
    """Event validation failed (bad event_type, severity, etc.)."""


@dataclass
class SecurityEvent:
    """A security event to be sent to LBRO.

    Args:
        event_type: One of the allowed event type strings.
        severity: critical / high / medium / low / info.
        source_ip: IP address of the event source.
        source_host: Hostname of the source machine.
        source_application: Application name that generated this event.
        message: Human-readable description of the event.
        event_timestamp: When the event occurred (UTC). Defaults to now.
        payload: Arbitrary JSON-serialisable key-value pairs.
    """
    event_type: str
    severity: str = "medium"
    source_ip: str | None = None
    source_host: str | None = None
    source_application: str | None = None
    message: str | None = None
    event_timestamp: str | None = None  # ISO 8601 UTC
    payload: dict[str, Any] = field(default_factory=dict)

    def __post_init__(self) -> None:
        if self.event_type not in EVENT_TYPES:
            raise LBROValidationError(
                f"Unknown event_type '{self.event_type}'. Allowed: {sorted(EVENT_TYPES)}"
            )
        if self.severity not in SEVERITIES:
            raise LBROValidationError(
                f"Unknown severity '{self.severity}'. Allowed: {sorted(SEVERITIES)}"
            )
        if self.event_timestamp is None:
            self.event_timestamp = datetime.now(timezone.utc).isoformat()

    def to_dict(self) -> dict:
        d = asdict(self)
        return {k: v for k, v in d.items() if v is not None}


class LBROClient:
    """LBRO event ingestion client.

    Args:
        api_key: Project API key (format: proj_<random>).
        base_url: Base URL of your LBRO instance (no trailing slash).
        timeout: HTTP request timeout in seconds. Default: 10.
        max_retries: Number of retries on transient errors. Default: 3.
        retry_delay: Base delay between retries in seconds. Default: 1.
        source_application: Default application name for all events.
        verify_ssl: Whether to verify SSL certificates. Default: True.
    """

    def __init__(
        self,
        api_key: str,
        base_url: str = "http://localhost:8000",
        timeout: float = 10.0,
        max_retries: int = 3,
        retry_delay: float = 1.0,
        source_application: str | None = None,
        verify_ssl: bool = True,
    ) -> None:
        if not api_key.startswith("proj_"):
            raise LBROAuthError("LBRO project API keys must start with 'proj_'")
        self._api_key = api_key
        self._base_url = base_url.rstrip("/")
        self._timeout = timeout
        self._max_retries = max_retries
        self._retry_delay = retry_delay
        self._source_application = source_application
        self._verify_ssl = verify_ssl

    # ── Public methods ──────────────────────────────────────────────────────

    def send_event(
        self,
        event_type: str,
        severity: str = "medium",
        source_ip: str | None = None,
        source_host: str | None = None,
        message: str | None = None,
        payload: dict[str, Any] | None = None,
        event_timestamp: str | None = None,
    ) -> dict:
        """Send a single security event. Returns the LBRO response dict."""
        event = SecurityEvent(
            event_type=event_type,
            severity=severity,
            source_ip=source_ip,
            source_host=source_host,
            source_application=self._source_application,
            message=message,
            event_timestamp=event_timestamp,
            payload=payload or {},
        )
        return self._post("/api/v1/events", event.to_dict())

    def send_events(self, events: list[SecurityEvent]) -> dict:
        """Send a batch of events. Returns summary with accepted/rejected counts."""
        if not events:
            return {"accepted": 0, "rejected": 0, "events": [], "errors": []}
        if len(events) > 1000:
            raise LBROValidationError("Batch size cannot exceed 1000 events")

        # Inject source_application default if not set per event
        dicts = []
        for e in events:
            d = e.to_dict()
            if "source_application" not in d and self._source_application:
                d["source_application"] = self._source_application
            dicts.append(d)

        return self._post("/api/v1/events/batch", {"events": dicts})

    def ping(self) -> bool:
        """Check connectivity to LBRO. Returns True if reachable."""
        try:
            self._get("/health")
            return True
        except Exception:
            return False

    # ── Internal HTTP helpers ───────────────────────────────────────────────

    def _post(self, path: str, body: dict) -> dict:
        return self._request("POST", path, body)

    def _get(self, path: str) -> dict:
        return self._request("GET", path)

    def _request(self, method: str, path: str, body: dict | None = None) -> dict:
        url = self._base_url + path
        data = json.dumps(body).encode("utf-8") if body is not None else None
        headers = {
            "Authorization": f"Bearer {self._api_key}",
            "Content-Type": "application/json",
            "Accept": "application/json",
            "User-Agent": f"lbro-python-sdk/{__version__}",
        }

        last_exc: Exception | None = None
        for attempt in range(self._max_retries + 1):
            try:
                req = URLRequest(url, data=data, headers=headers, method=method)
                with urlopen(req, timeout=self._timeout) as resp:
                    body_bytes = resp.read()
                    return json.loads(body_bytes) if body_bytes else {}

            except HTTPError as exc:
                status = exc.code
                body_text = exc.read().decode("utf-8", errors="replace")
                if status == 401:
                    raise LBROAuthError(f"Invalid project API key: {body_text}") from exc
                if status == 422:
                    raise LBROValidationError(f"Validation error: {body_text}") from exc
                if status >= 500:
                    last_exc = exc
                    if attempt < self._max_retries:
                        delay = self._retry_delay * (2 ** attempt)
                        log.warning("LBRO request failed (HTTP %s), retrying in %.1fs", status, delay)
                        time.sleep(delay)
                        continue
                raise LBROError(f"HTTP {status}: {body_text}") from exc

            except (URLError, OSError) as exc:
                last_exc = exc
                if attempt < self._max_retries:
                    delay = self._retry_delay * (2 ** attempt)
                    log.warning("LBRO network error: %s — retrying in %.1fs", exc, delay)
                    time.sleep(delay)
                    continue

        raise LBROError(f"Request failed after {self._max_retries} retries: {last_exc}") from last_exc

File: sdk/nodejs/lbro-sdk.js
/**
 * LBRO Node.js SDK
 *
 * Lightweight SDK for sending security events to LBRO from Node.js applications.
 *
 * @example
 * const { LBROClient } = require('./lbro-sdk');
 *
 * const client = new LBROClient({
 *   apiKey: 'proj_your_project_api_key',
 *   baseUrl: 'https://your-lbro-instance.example.com',
 *   sourceApplication: 'my-node-app',
 * });
 *
 * // Single event
 * await client.sendEvent({
 *   eventType: 'auth_failure',
 *   severity: 'high',
 *   sourceIp: '203.0.113.42',
 *   message: 'Failed login attempt',
 *   payload: { username: 'admin', attempts: 10 },
 * });
 *
 * // Batch
 * await client.sendEvents([
 *   { eventType: 'sql_injection', severity: 'critical', message: 'SQL injection in /search' },
 *   { eventType: 'xss', severity: 'medium', message: 'XSS in comment field' },
 * ]);
 */
'use strict';

const https = require('https');
const http = require('http');
const { URL } = require('url');

const SDK_VERSION = '1.0.0';

const ALLOWED_EVENT_TYPES = new Set([
  'auth_failure', 'sql_injection', 'xss', 'brute_force', 'port_scan',
  'suspicious_request', 'system_log', 'application_log', 'nginx_log',
  'apache_log', 'firewall_event', 'windows_event', 'linux_audit', 'custom',
]);

const ALLOWED_SEVERITIES = new Set(['critical', 'high', 'medium', 'low', 'info']);

class LBROError extends Error {
  constructor(message, statusCode = null) {
    super(message);
    this.name = 'LBROError';
    this.statusCode = statusCode;
  }
}

class LBROAuthError extends LBROError {
  constructor(message) {
    super(message, 401);
    this.name = 'LBROAuthError';
  }
}

class LBROValidationError extends LBROError {
  constructor(message) {
    super(message, 422);
    this.name = 'LBROValidationError';
  }
}

class LBROClient {
  /**
   * @param {object} options
   * @param {string} options.apiKey          - Project API key (starts with proj_)
   * @param {string} [options.baseUrl]       - LBRO base URL. Default: http://localhost:8000
   * @param {number} [options.timeout]       - Request timeout ms. Default: 10000
   * @param {number} [options.maxRetries]    - Max retries on 5xx/network error. Default: 3
   * @param {number} [options.retryDelay]    - Base retry delay ms. Default: 1000
   * @param {string} [options.sourceApplication] - Default app name for all events
   */
  constructor({ apiKey, baseUrl = 'http://localhost:8000', timeout = 10000,
                maxRetries = 3, retryDelay = 1000, sourceApplication = null } = {}) {
    if (!apiKey || !apiKey.startsWith('proj_')) {
      throw new LBROAuthError("LBRO project API keys must start with 'proj_'");
    }
    this._apiKey = apiKey;
    this._baseUrl = baseUrl.replace(/\/$/, '');
    this._timeout = timeout;
    this._maxRetries = maxRetries;
    this._retryDelay = retryDelay;
    this._sourceApplication = sourceApplication;
  }

  /**
   * Send a single security event.
   * @param {object} event
   * @returns {Promise<object>} LBRO response
   */
  async sendEvent({ eventType, severity = 'medium', sourceIp = null, sourceHost = null,
                    message = null, payload = {}, eventTimestamp = null } = {}) {
    this._validateEventType(eventType);
    this._validateSeverity(severity);
    const body = this._buildEventBody({ eventType, severity, sourceIp, sourceHost,
                                         message, payload, eventTimestamp });
    return this._post('/api/v1/events', body);
  }

  /**
   * Send multiple events in a single batch request (max 1000).
   * @param {object[]} events
   * @returns {Promise<object>} Summary with accepted/rejected counts
   */
  async sendEvents(events) {
    if (!Array.isArray(events) || events.length === 0) {
      return { accepted: 0, rejected: 0, events: [], errors: [] };
    }
    if (events.length > 1000) {
      throw new LBROValidationError('Batch size cannot exceed 1000 events');
    }
    const bodies = events.map(e => {
      this._validateEventType(e.eventType);
      this._validateSeverity(e.severity || 'medium');
      return this._buildEventBody(e);
    });
    return this._post('/api/v1/events/batch', { events: bodies });
  }

  /** Check connectivity. Returns true if LBRO is reachable. */
  async ping() {
    try {
      await this._get('/health');
      return true;
    } catch (_) {
      return false;
    }
  }

  // ── Private ────────────────────────────────────────────────────────────────

  _buildEventBody({ eventType, severity = 'medium', sourceIp, sourceHost,
                    message, payload = {}, eventTimestamp } = {}) {
    const body = {
      event_type: eventType,
      severity,
      payload,
      event_timestamp: eventTimestamp || new Date().toISOString(),
    };
    if (sourceIp)          body.source_ip = sourceIp;
    if (sourceHost)        body.source_host = sourceHost;
    if (message)           body.message = message;
    if (this._sourceApplication) body.source_application = this._sourceApplication;
    return body;
  }

  _validateEventType(eventType) {
    if (!ALLOWED_EVENT_TYPES.has(eventType)) {
      throw new LBROValidationError(
        `Unknown eventType '${eventType}'. Allowed: ${[...ALLOWED_EVENT_TYPES].sort().join(', ')}`
      );
    }
  }

  _validateSeverity(severity) {
    if (!ALLOWED_SEVERITIES.has(severity)) {
      throw new LBROValidationError(
        `Unknown severity '${severity}'. Allowed: ${[...ALLOWED_SEVERITIES].sort().join(', ')}`
      );
    }
  }

  async _post(path, body) {
    return this._request('POST', path, body);
  }

  async _get(path) {
    return this._request('GET', path);
  }

  async _request(method, path, body = null) {
    const url = new URL(this._baseUrl + path);
    const isHttps = url.protocol === 'https:';
    const transport = isHttps ? https : http;
    const bodyStr = body ? JSON.stringify(body) : null;

    const options = {
      hostname: url.hostname,
      port: url.port || (isHttps ? 443 : 80),
      path: url.pathname + url.search,
      method,
      headers: {
        Authorization: `Bearer ${this._apiKey}`,
        'Content-Type': 'application/json',
        Accept: 'application/json',
        'User-Agent': `lbro-nodejs-sdk/${SDK_VERSION}`,
        ...(bodyStr ? { 'Content-Length': Buffer.byteLength(bodyStr) } : {}),
      },
    };

    let lastError = null;
    for (let attempt = 0; attempt <= this._maxRetries; attempt++) {
      try {
        return await new Promise((resolve, reject) => {
          const req = transport.request(options, (res) => {
            let data = '';
            res.on('data', chunk => { data += chunk; });
            res.on('end', () => {
              const status = res.statusCode;
              if (status === 401) return reject(new LBROAuthError(`Invalid project API key: ${data}`));
              if (status === 422) return reject(new LBROValidationError(`Validation error: ${data}`));
              if (status >= 500) return reject(new LBROError(`Server error HTTP ${status}: ${data}`, status));
              if (status >= 400) return reject(new LBROError(`HTTP ${status}: ${data}`, status));
              try { resolve(data ? JSON.parse(data) : {}); }
              catch (e) { resolve({}); }
            });
          });
          req.setTimeout(this._timeout, () => {
            req.destroy();
            reject(new LBROError(`Request timeout after ${this._timeout}ms`));
          });
          req.on('error', reject);
          if (bodyStr) req.write(bodyStr);
          req.end();
        });
      } catch (err) {
        lastError = err;
        if (err instanceof LBROAuthError || err instanceof LBROValidationError) throw err;
        if (attempt < this._maxRetries) {
          const delay = this._retryDelay * Math.pow(2, attempt);
          await new Promise(r => setTimeout(r, delay));
          continue;
        }
      }
    }
    throw new LBROError(`Request failed after ${this._maxRetries} retries: ${lastError?.message}`, null);
  }
}

module.exports = { LBROClient, LBROError, LBROAuthError, LBROValidationError, ALLOWED_EVENT_TYPES, ALLOWED_SEVERITIES };
